### PR TITLE
[Feature](array-function) Support array_concat function

### LINKED ISSUE #17307: [Feature](array function) Support `array_concat` function
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description
Concat all arrays passed in the arguments.
```
mysql> select array_concat([1, 2], [7, 8], [5, 6]);
+-----------------------------------------------------+
| array_concat(ARRAY(1, 2), ARRAY(7, 8), ARRAY(5, 6)) |
+-----------------------------------------------------+
| [1, 2, 7, 8, 5, 6]                                  |
+-----------------------------------------------------+
```

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by gitccl**
# Proposed changes

Issue Number: close #17307 

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [X] Has document been added or modified
* [ ] Does it need to update dependencies
* [X] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by gitccl** [comment]
run buildall

File: be/src/vec/functions/array/function_array_concat.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (eldenmoon) [inline be/src/vec/functions/array/function_array_concat.cpp:1]: also need add this doc to sidebar `docs/sidebars.json`

File: regression-test/suites/query_p0/sql_functions/array_functions/test_array_functions_by_literal.groovy [hunk 2]
```diff
@@ -223,6 +223,14 @@ suite("test_array_functions_by_literal") {
     qt_sql """select array_apply(array(cast (24.99 as decimal(10,3)),cast (25.99 as decimal(10,3))), ">", '25')"""
     qt_sql """select array_apply(array(cast (24.99 as decimal(10,3)),cast (25.99 as decimal(10,3))), "!=", '25.99')"""
 
+    qt_sql "select array_concat([1, 2, 3], [2, 3, 4], [8, 1, 2], [9])"
+    qt_sql "select array_concat([12, 23], [25, null], [null], [66])"
+    qt_sql "select array_concat([1.2, 1.8], [9.0, 2.2], [2.8])"
+    qt_sql "select array_concat(['aaa', null], ['bbb', 'fff'], [null, 'ccc'])"
```
Comment (eldenmoon) [inline regression-test/suites/query_p0/sql_functions/array_functions/test_array_functions_by_literal.groovy:229]: add a null case like array_concat(null, [1, 2, 3],  null)

File: be/src/vec/functions/array/function_array_concat.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return arguments[0];
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        const size_t result, size_t input_rows_count) override {
+        DataTypePtr column_type = block.get_by_position(arguments[0]).type;
+        auto nested_type = assert_cast<const DataTypeArray&>(*column_type).get_nested_type();
+        auto result_column = ColumnArray::create(nested_type->create_column(),
+                                                 ColumnArray::ColumnOffsets::create());
+        IColumn& result_nested_col = result_column->get_data();
```
Comment (eldenmoon) [inline be/src/vec/functions/array/function_array_concat.cpp:56]: we could do `reserve` for result_nested_col for more efficiency

**Comment by gitccl** [comment]
run buildall

File: be/src/vec/functions/array/function_array_concat.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        ColumnArray::Offsets64& column_offsets = result_column->get_offsets();
+        column_offsets.resize(input_rows_count);
+
+        size_t total_size = 0;
+        for (size_t col : arguments) {
+            ColumnPtr src_column =
+                    block.get_by_position(col).column->convert_to_full_column_if_const();
+            const auto& src_column_array = check_and_get_column<ColumnArray>(*src_column);
```
Comment (eldenmoon) [inline be/src/vec/functions/array/function_array_concat.cpp:64]: we should get the nested column of array, and caculate it's size, other wise, it's the row size of array not the inner nested row size of the array.

**Review by eldenmoon — COMMENTED** [review]
array size is not precise to do reserve

File: be/src/vec/functions/array/function_array_concat.cpp (hunk 4 above)
Comment (gitccl) [reply be/src/vec/functions/array/function_array_concat.cpp:64]: done

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Comment by eldenmoon** [comment]
run beut

**Comment by eldenmoon** [comment]
run feut

**Comment by eldenmoon** [comment]
This pr has conflicts, please rebase

**Comment by gitccl** [comment]
> This pr has conflicts, please rebase

done

**Review by eldenmoon — APPROVED** [review]
LGTM

**Comment by eldenmoon** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**